### PR TITLE
Dokken/fixissue157

### PR DESCRIPTION
**Pull request by jorgensd**
Removing the local hack for finding number of dofs for a coordinate_element component, using fiat interface.

### CODE REVIEW

File: ffc/codegeneration/definitions.py [hunk 1]
```diff
@@ -16,34 +16,15 @@
...
-    The table data is to messy to figure out a clean design for that quickly.
     """
-    from ufl.cell import num_cell_entities
-    degree = coordinate_element.degree()
-    cell = coordinate_element.cell()
-    tdim = cell.topological_dimension()
-    cellname = cell.cellname()
+    from ffc.fiatinterface import create_element
+    # Extracting only first component degrees of freedom from FIAT
+    fiat_element = create_element(coordinate_element).elements()[0]
```
Comment (michalhabera) [inline ffc/codegeneration/definitions.py]: Please assert that all components have the same element.

File: ffc/codegeneration/definitions.py [hunk 2]
```diff
@@ -16,34 +16,18 @@
...
-            else:
-                n = (degree - 2) * (degree - 1) // 2
-        elif entity_dim == 3:
-            n = (degree - 3) * (degree - 2) * (degree - 1) // 6
-        else:
-            raise RuntimeError("Entity dimension out of range")
-        # Accumulate
-        num_entities = num_cell_entities[cellname][entity_dim]
-        d += num_entities * n
+    for entity_dim in dof_dict.keys():
```
Comment (michalhabera) [inline ffc/codegeneration/definitions.py]: What about using
```
    def space_dimension(self):
        """Return the dimension of the finite element space."""
        return len(self.dual_basis())
```
?

`entity_dofs` doesnt return e.g. dofs of DG elements, which live in the interior of the cell, you'd need to use `entity_closure_dofs`. I know coordinate map is not discontinuous, but just for clarity.

Comment (jorgensd) [reply ffc/codegeneration/definitions.py]: Sure, seems ok. Reason for the other version was to extend the old hack to arbitrary lagrangian elements.

Comment (michalhabera) [reply ffc/codegeneration/definitions.py]: One more comment. You can notice currently, that FIAT is interfaced only in `fiatiinterface.py`. Can you refactor a but and query FIAT library only in `fiatinterface.py`?

Comment (jorgensd) [reply ffc/codegeneration/definitions.py]: Is that true? I can see it accessed in: codegeneration/access.py

Comment (michalhabera) [reply ffc/codegeneration/definitions.py]: Ah, sorry. I thought you were importing FIAT library, its fine like this, nvm.

**Comment by michalhabera** [comment]
@jorgensd Do you have a dolfinx branch to test against?

File: ffc/codegeneration/definitions.py [hunk 3]
```diff
@@ -16,35 +16,13 @@
...
-                n = (degree - 2) * (degree - 1) // 2
-        elif entity_dim == 3:
-            n = (degree - 3) * (degree - 2) * (degree - 1) // 6
-        else:
-            raise RuntimeError("Entity dimension out of range")
-        # Accumulate
-        num_entities = num_cell_entities[cellname][entity_dim]
-        d += num_entities * n
-    return d
+    from ffc.fiatinterface import create_element
```
Comment (chrisrichardson) [inline ffc/codegeneration/definitions.py]: I thought flake8 was supposed to stop you importing in the middle of a file...?

Comment (jorgensd) [reply ffc/codegeneration/definitions.py]: Good question. Have moved it regardless

**Comment by jorgensd** [comment]
@michalhabera https://github.com/FEniCS/dolfinx/tree/dokken/ffc-fixissue157

**Pull request merged**